Assembler.UnitTests/InitialMessageHandlerTests.cs

```diff
@@ -0,0 +1,182 @@
+using Assembler.Base;
+using Assembler.Core;
+using Assembler.Core.Entities;
+using Assembler.Core.Enums;
+using Moq;
+using NUnit.Framework;
+
+namespace Assembler.UnitTests
+{
+    [TestFixture]
+    public class InitialMessageHandlerTests
+    {
+        private Mock<ITimeBasedCache<BaseMessageInAssembly>> _cacheMock;
+        private Mock<IFactory<BaseFrame, string>> _identifierFactoryMock;
+        private Mock<IMessageEnricher<BaseFrame, BaseMessageInAssembly>> _enricherMock;
+        private Mock<ICreator<BaseMessageInAssembly>> _messageInAssemblyCreatorMock;
+
+        private List<BaseMessageInAssembly> _assembledMessages;
+        private string _identifierString;
+
+        private InitialMessageHandler<BaseFrame, BaseMessageInAssembly> _handler;
+
+        [SetUp]
+        public void Setup()
+        {
+            _cacheMock = new Mock<ITimeBasedCache<BaseMessageInAssembly>>();
+            _enricherMock = new Mock<IMessageEnricher<BaseFrame, BaseMessageInAssembly>>();
+            _messageInAssemblyCreatorMock = new Mock<ICreator<BaseMessageInAssembly>>();
+
+            _identifierString = Utilities.GetIdentifierString();
+            _identifierFactoryMock = Utilities.GetIdentifierMock();
+
+            _assembledMessages = new List<BaseMessageInAssembly>();
+
+            _handler = new InitialMessageHandler<BaseFrame, BaseMessageInAssembly>(_cacheMock.Object,
+                _identifierFactoryMock.Object, _messageInAssemblyCreatorMock.Object, _enricherMock.Object,
+                Utilities.GetLoggerFactory());
+
+            _handler.MessageAssemblyFinished += _assembledMessages.Add;
+        }
+
+        [TearDown]
+        public void Teardown()
+        {
+            _identifierFactoryMock.VerifyNoOtherCalls();
+            _enricherMock.VerifyNoOtherCalls();
+            _cacheMock.VerifyNoOtherCalls();
+            _messageInAssemblyCreatorMock.VerifyNoOtherCalls();
+        }
+
+        [Test]
+        public void Handle_IdentifierThrowsException_FrameNotBeingUsed()
+        {
+            // Arrange
+            var frame = new Mock<BaseFrame>(FrameType.Initial);
+
+            _identifierFactoryMock.Setup(identifier => identifier.Create(It.IsAny<BaseFrame>()))
+                .Throws<NullReferenceException>();
+
+            // Act
+            _handler.Handle(frame.Object);
+
+            // Assert
+            _identifierFactoryMock.Verify(identifier => identifier.Create(It.IsAny<BaseFrame>()), Times.Once);
+            _identifierFactoryMock.Verify(identifier => identifier.Create(frame.Object), Times.Once);
+        }
+
+        [Test]
+        public void Handle_MessageInCacheAndMiddleReceived_ReleasesTheOldOneAndCreatesANewOne()
+        {
+            // Arrange
+            var frame = new Mock<BaseFrame>(FrameType.Initial);
+            var message = new Mock<BaseMessageInAssembly>();
+            var newMessage = new Mock<BaseMessageInAssembly>();
+            message.Object.MiddleReceived = true;
```

Comment:
`message.SetupGet(msg => msg.MiddleReceived).Returns(true);`

Reply:
Throws exception, I used it before but it wouldn't let me.

Reply:
As it seems, that's because:
a. It's a field, not a property (because no `{ get; set; }`)
b. Even if it's a property, it needs to be overridable (virtual/abstract)

Well, you learn new things everyday I guess
For now you can keep it like this if you don't wanna quarrel with it too much